doc/lsst.daf.butler/formatters.rst

```diff
@@ -0,0 +1,316 @@
+.. _daf_butler_storageclass_formatters_assemblers:
+
+##########################################
+Storage Classes, Assemblers and Formatters
+##########################################
+
+Formatters and assemblers provide the interface between Butler and the python types it is storing and retrieving.
+A Formatter is responsible for serializing a Python type to an external storage system and reading that serialized form back into Python.
+The serialization can be stored to a local file system or cloud storage or even a database.
+It is possible for a formatter to be globally configured to use particular parameters on write.
+On retrieval of datasets read parameters can be used that can, for example, return only a subset of the data.
+
+Assemblers are used to disassemble and reassemble composite datasets and can also be used to process read parameters that adjust how the retrieved dataset might be modified on get.
+
+Deciding which formatter or assembler to use is controlled by the storage class and corresponding dataset type.
+
+Storage Classes
+===============
+
+A Storage Class is fundamental to informing Butler how to deal with specific Python types.
+Each `~lsst.daf.butler.DatasetType` is associated with a `~lsst.daf.butler.StorageClass`.
+This storage class declares the Python type, any components it may have (read-only or read-write), and an assembler that can be used to process read parameters.
+
+Composites
+^^^^^^^^^^
+
+A composite storage class declares that the Python type consists of discrete components that can be accessed individually.
+Each of these components must declare its storage class as well.
+
+For example, if a ``pvi`` dataset type has been associated with an ``ExposureF`` storage class, the user can `~lsst.daf.butler.Butler.get()` the full ``pvi`` and access the components as they would normally for an `~lsst.afw.image.ExposureF`, or if the user solely want the metadata header from the exposure they can ask for ``pvi.metadata`` and just get that.
```

Comment:
Ah, so `ExposureF` is a "composite storage class"?

Reply:
Yes.